src/frontend/src/planner/select.rs

```diff
@@ -413,6 +428,139 @@ impl Planner {
+            correlated_id: Option<CorrelatedId>,
+            correlated_indices_collection: Vec<Vec<usize>>,
+            ctx: OptimizerContextRef,
+        }
+
+        impl ExprRewriter for SubstituteSubQueries {
+            fn rewrite_subquery(&mut self, mut subquery: Subquery) -> ExprImpl {
+                if self.correlated_id.is_none() {
+                    self.correlated_id = Some(self.ctx.next_correlated_id());
+                }
+                let input_ref = InputRef::new(self.input_col_num, subquery.return_type()).into();
+                self.input_col_num += 1;
+                self.correlated_indices_collection.push(
+                    subquery.collect_correlated_indices_by_depth_and_assign_id(
+                        0,
+                        self.correlated_id.unwrap(),
+                    ),
+                );
+                self.subqueries.push(subquery);
+                input_ref
+            }
+        }
+
+        let mut rewriter = SubstituteSubQueries {
+            input_col_num: root.schema().len(),
+            subqueries: vec![],
+            correlated_id: None,
+            correlated_indices_collection: vec![],
+            ctx: self.ctx.clone(),
+        };
+        exprs = exprs
+            .into_iter()
+            .map(|e| rewriter.rewrite_expr(e))
+            .collect();
+
+        let mut right = None;
+
+        for subquery in rewriter.subqueries {
+            let return_type = subquery.return_type();
+            let subroot = self.plan_query(subquery.query)?;
+
+            let subplan = match subquery.kind {
+                SubqueryKind::Scalar => subroot.into_unordered_subplan(),
+                SubqueryKind::UpdateSet => {
+                    let plan = subroot.into_unordered_subplan();
+
+                    // Compose all input columns into a struct with `ROW` function.
+                    let all_input_refs = plan
+                        .schema()
+                        .data_types()
+                        .into_iter()
+                        .enumerate()
+                        .map(|(i, data_type)| InputRef::new(i, data_type).into())
+                        .collect::<Vec<_>>();
+                    let call =
+                        FunctionCall::new_unchecked(ExprType::Row, all_input_refs, return_type);
+
+                    LogicalProject::create(plan, vec![call.into()])
+                }
+                SubqueryKind::Existential => {
+                    self.create_exists(subroot.into_unordered_subplan())?
+                }
+                SubqueryKind::Array => subroot.into_array_agg()?,
+                _ => bail_not_implemented!(issue = 1343, "{:?}", subquery.kind),
+            };
+            if right.is_none() {
+                right = Some(subplan);
+            } else {
+                right = Some(LogicalJoin::create(
+                    right.clone().unwrap(),
+                    subplan,
+                    JoinType::Inner,
```

Comment:
For scalar subquery, the result may be 0 rows and will be replaced by 1 row of null. This can be problematic in such inner cross join:
```
create table a (a1 int);
create table b (b1 int);
insert into b values (2);
select 1+2, (select a1 from a);
select 1+2, (select b1 from b);
select 1+2, (select a1 from a), (select b1 from b); -- expect (3,null,2) actual (3,null,null)
```